docs/codelabs/codelab-chat-with-a-pdf.md

```diff
@@ -0,0 +1,206 @@
+
+
+Create a new folder `src`, and iniside it, a new file `index.ts`. Add the following lines to import Genkit and the Google AI plugin:
+
+```typescript
+import {gemini15Flash, googleAI} from '@genkit-ai/googleai';
+import {genkit} from 'genkit';
+```
+
+
+Add the following lines to configure Genkit and set Gemini 1.5 Flash as the default model:
+
+```typescript
+const ai = genkit({
+  plugins: [googleAI()],
+  model: gemini15Flash,
+});
+```
+
+
+Add the main body of your app:
+
+```typescript
+(async () => {
+  try {
+    // 1: get command line arguments
+    // 2: load PDF file
+    // 3: construct prompt
+    // 4: start chat
+    // 5: chat loop
+  } catch (error) {
+    console.error("Error parsing PDF or interacting with Genkit:", error);
+  }
+})(); // <-- don't forget the trailing parentesis to call the function!
+```
+
+
+## Load and parse a PDF file
+
+
+In this step, you will write code to load and parse a PDF file. PDF files can be quite complex, and parsing them is notoriously difficult, so making sure any type of PDF file can be parse is outside the scope of this codelab. The PDF parser you will be using should do a reasonably good job on most PDF files.
+
+First, install `pdf-parse`:
+
+```typescript
+npm i pdf-parse
+```
+
+
+Import the PDF libray to your app
+
+```typescript
+import pdf from 'pdf-parse';
+import fs from 'fs';
+```
+
+
+Read the name of the PDF file from the command line:
+
+```typescript
+	// 1: get command line arguments
+	const filename = process.argv[2];
+	if (!filename) {
+		console.error("Please provide a filename as a command line argument.");
+		process.exit(1);
+	}
+```
+
+
+Load the contents of the PDF file
+
+```typescript
+	// 2: load PDF file
+	let dataBuffer = fs.readFileSync(filename);
+	const { text } = await pdf(dataBuffer);
+```
+
+
+## Set up the prompt
+
+
+Allow the user to provide a custom prompt via the command line. If they don’t provide a prompt, use a default:
+
+```typescript
+const prefix = process.argv[3] || "Answer the user's questions about the contents of this PDF file.";
+```
+
+
+Inject the prompt prefix and the full text of the PDF file into the prompt for the model:
+
+```typescript
+    const prompt = `
+      ${prefix}
+      Context:
+      ${data.text}
+    `
+```
+
+
+## Implement the chat loop
+
+
+Start the chat with the model by calling the `chat` method, passing the prompt (which includes the full text of the PDF file):
+
+```typescript
+const chat = ai.chat({ system: prompt })
+```
+
+
+Import `createInterface` - this will allow you to build a text-based UI:
+
+```typescript
+import {createInterface} from "node:readline/promises";
+```
+
+
+Instantiate a text input, then display a message to the user:
+
+```typescript
+    const readline = createInterface(process.stdin, process.stdout);
+    console.log("You're chatting with Gemini. Ctrl-C to quit.\n");
+```
+
+
+In an endless loop, read the user’s input, then send it to the model using `chat.send`:
```

Comment:
What will end the loop? Maybe rewrite as:

"Read the user's input...send`. This operation loops until <end condition>."